compiler_opt/es/regalloc_trace/regalloc_trace_worker.py

```diff
@@ -64,6 +74,12 @@ def __init__(self, clang_path: str, basic_block_trace_model_path: str,
     self._thread_count = thread_count
     self._corpus_path = corpus_path
 
+    gin.parse_config(gin_config)
+    self._setup_base_policy()
+
+  def __del__(self):
+    shutil.rmtree(self._tf_base_temp_dir)
```

Comment:
iiuc `__del__` happens at GC. Can we do this deterministically, e.g. based on a scope?

Reply:
Python doesn't really have the notion of running a destructor after falling out of scope as far as I can tell, so this is sort of the closest we can get.

We could set up the runner as a context manager, but instantiating each runner with a `with` statement won't work.

Reply:
Right, my point is that the dtor will be run non-deterministically, at GC.

Why is `with` not an option for each worker? Because it'd add an extra responsibility?

Oh, but this is a worker-lifetime bound tempdir, correct? so it's a bit of "best effort", right?

Reply:
Yeah, it is bound by the lifetime of the worker. If it leaks it isn't a big deal, but not ideal.

GIven it's in the constructor of the worker, we'd have to setup the worker as a context manager which is incompatible with how we have worker pools setup. The other approach would be to have the worker pool call some close method or something that deletes the temp dir/cleans up, but there will still be issues with that where things don't get cleaned up, so I think this is a reasonable enough approach.

Reply:
add a comment that this is best effort, as it happens at shutdown and, if the shutdown is forced (not uncommon on cloud setups), :shrug: (and probably doesn't matter since there's more aggressive cleanup happening there anyway)